### ISSUE TITLE
Error on login: NoneType object has no attribute 'cursor'

### ISSUE THREAD

**Issue by varac** [pixelated/pixelated-user-agent#773]
On unstable, we just noticed that the useragent throws multiple errors like this:

```
2016-08-30T11:35:28.179009+02:00 unstable1 pixelated-user-agent[25663]: 2016-08-30 11:35:28 [twisted] CRITICAL Unhandled Error
2016-08-30T11:35:28.179015+02:00 unstable1 pixelated-user-agent[25663]: Traceback (most recent call last):
2016-08-30T11:35:28.179021+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 434, in errback
2016-08-30T11:35:28.179027+02:00 unstable1 pixelated-user-agent[25663]: self._startRunCallbacks(fail)
2016-08-30T11:35:28.179032+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 501, in _startRunCallbacks
2016-08-30T11:35:28.179039+02:00 unstable1 pixelated-user-agent[25663]: self._runCallbacks()
2016-08-30T11:35:28.179044+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 588, in _runCallbacks
2016-08-30T11:35:28.179051+02:00 unstable1 pixelated-user-agent[25663]: current.result = callback(current.result, *args, **kw)
2016-08-30T11:35:28.179056+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1184, in gotResult
2016-08-30T11:35:28.179062+02:00 unstable1 pixelated-user-agent[25663]: _inlineCallbacks(r, g, deferred)
2016-08-30T11:35:28.179068+02:00 unstable1 pixelated-user-agent[25663]: --- <exception caught here> ---
2016-08-30T11:35:28.179074+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
2016-08-30T11:35:28.179080+02:00 unstable1 pixelated-user-agent[25663]: result = result.throwExceptionIntoGenerator(g)
2016-08-30T11:35:28.179086+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
2016-08-30T11:35:28.179092+02:00 unstable1 pixelated-user-agent[25663]: return g.throw(self.type, self.value, self.tb)
2016-08-30T11:35:28.179097+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/src/leap.soledad.client/client/src/leap/soledad/client/sqlcipher.py", line 514, in sync
2016-08-30T11:35:28.179103+02:00 unstable1 pixelated-user-agent[25663]: defer_decryption=defer_decryption)
2016-08-30T11:35:28.179108+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1126, in _inlineCallbacks
2016-08-30T11:35:28.179114+02:00 unstable1 pixelated-user-agent[25663]: result = result.throwExceptionIntoGenerator(g)
2016-08-30T11:35:28.179120+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/local/lib/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
2016-08-30T11:35:28.179126+02:00 unstable1 pixelated-user-agent[25663]: return g.throw(self.type, self.value, self.tb)
2016-08-30T11:35:28.179135+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/src/leap.soledad.client/client/src/leap/soledad/client/sync.py", line 136, in sync
2016-08-30T11:35:28.179141+02:00 unstable1 pixelated-user-agent[25663]: my_gen, _, changes = self.source.whats_changed(target_my_gen)
2016-08-30T11:35:28.179147+02:00 unstable1 pixelated-user-agent[25663]: File "/usr/share/python/pixelated-user-agent/src/leap.soledad.common/common/src/leap/soledad/common/l2db/backends/sqlite_backend.py", line 418, in whats_changed
2016-08-30T11:35:28.179153+02:00 unstable1 pixelated-user-agent[25663]: c = self._db_handle.cursor()
2016-08-30T11:35:28.179159+02:00 unstable1 pixelated-user-agent[25663]: exceptions.AttributeError: 'NoneType' object has no attribute 'cursor'
```

**Comment by SweetVirginia** [comment]
@varac Could you explain that to me? What are those errors?

**Comment by varac** [comment]
This is from the  pixelated-server daemon logfile (`/var/log/pixelated/user-agent.log`) on the server (in this case unstable.pix).
To be honest, i don't have any clue what it's about. We found this while debugging the 5mb attachment issue, but it is not related to it, because it happens way before during login/sync.
I think someone should have a deeper look into what's happening here.

**Comment by bwagnerr** [comment]
Hey, just as a heads up, a quick update. We have this problem happening in almost all environments now, this was coupled with another problem (KeyError on logout) and I think they might be related.

I fixed the other problem first with the commit https://github.com/pixelated/pixelated-user-agent/commit/a274d4cf3628816d21bb9848f019a444119b80a5

Now we should wait for the deploy on the environments to see if this problem still persists, if it does, we'll prioritize this

**Comment by bwagnerr** [comment]
@mnandri managed to reproduce the issue [here](https://github.com/pixelated/puppet-pixelated/issues/53#issuecomment-248887036) and that problem is affecting our functional tests even after the previous fix.

I'll just reproduce his findings here:

> This is a failure during soledad sync. It happens when it fails to access the user sqlite file: see code.
> 
> python2 sqlite3 docs says the database is locked if one connection is changing it, until that transaction is committed or the 5s timeout is reached.
> possible culprit here: logout or soledad.close doesn't immediatly close all db connections.

> update:
> logout is actually causing the self._db_handle to be None.


We'll resume investigation on this issue

**Comment by mnandri** [comment]
The reason for this is because at each login we start a mail fetcher (see code) but on logout, we only stop only one of the fetchers (started by the other logins - separate sessions/browsers). Since we cache the leap_sessions, we only use one soledad, keymanager... instances and those are all closed on logout... in particular, db connections are closed (or rather set to None)... and hence the  `'NoneType' object has no attribute 'cursor'`

**Comment by thaissiqueira** [comment]
To reproduce the error:
- I did rollback for one commit before the correction
- I started pixelated-user-agent (with rollback mentioned above) on vagrant using dev as provider
- I did login with on Chrome
- I did login again, with the same user, on FireFox
- I did the logout on Chrome, then the user was logout on FireFox too
- Then I did login again on Chrome to start sync

So, the exception mentioned above as showed.

After that I did checkout to master and did the same steps mentioned above.
Then I could check that the exception doesn't happen again.

An @tuliocasagrande helped me with code review (thanks Tulio!)
